build_runner/lib/src/entrypoint/run_script.dart

```diff
@@ -0,0 +1,145 @@
+    }
+
+    var scriptName = argResults.rest[0];
+    var passedArgs = argResults.rest.skip(1).toList();
+    var options = readOptions();
+
+    // Create a temporary directory in which to execute the script.
+    var tempPath = Directory.systemTemp
+        .createTempSync('build_runner_run_script')
+        .absolute
+        .uri
+        .toFilePath();
+
+    // Create two ReceivePorts, so that we can quit when the isolate is done.
+    //
+    // Define these before starting the isolate, so that we can close
+    // them if there is a spawn exception.
+    ReceivePort onExit, onError;
+
+    try {
+      var outputMap = options.outputMap ?? {};
+      outputMap.addAll({tempPath: null});
+
+      var result = await build(
+        builderApplications,
+        deleteFilesByDefault: options.deleteFilesByDefault,
+        enableLowResourcesMode: options.enableLowResourcesMode,
+        configKey: options.configKey,
+        assumeTty: options.assumeTty,
+        outputMap: outputMap,
+        packageGraph: packageGraph,
+        verbose: options.verbose,
+        builderConfigOverrides: options.builderConfigOverrides,
+        isReleaseBuild: options.isReleaseBuild,
+        trackPerformance: options.trackPerformance,
+        skipBuildScriptCheck: options.skipBuildScriptCheck,
+        buildDirs: options.buildDirs,
+        logPerformanceDir: options.logPerformanceDir,
+      );
+
+      if (result.status == BuildStatus.failure) {
+        stdout.writeln('Skipping script run due to build failure');
+        return result.failureType.exitCode;
+      }
+
+      // Find the path of the script to run.
+      var scriptPath = p.setExtension(p.join(tempPath, scriptName), '.dart');
+      var packageConfigPath = p.join(tempPath, '.packages');
+
+      // Use a completer to determine the exit code.
+      var completer = new Completer<int>();
+
+      Isolate isolate;
+      onExit = new ReceivePort();
+      onError = new ReceivePort();
+
+      // On an error, kill the isolate, and rethrow the error.
+      onError.listen((e) {
+        isolate?.kill();
+        onExit.close();
+        onError.close();
+        Zone.current.handleUncaughtError(
+            e[0], new StackTrace.fromString(e[1].toString()));
+      });
+
+      onExit.listen((_) {
+        onExit.close();
+        onError.close();
+        if (!completer.isCompleted) {
+          completer.complete(ExitCode.success.code);
+        }
+      });
+
+      isolate = await Isolate.spawnUri(
+        p.toUri(scriptPath),
+        passedArgs,
+        null,
+        onExit: onExit.sendPort,
+        onError: onError.sendPort,
+        packageConfig: p.toUri(packageConfigPath),
+      );
+
+      return await completer.future;
+    } on IsolateSpawnException catch (e) {
+      stderr.writeln(e);
+      onExit?.close();
+      onError?.close();
+      stderr.writeln(
+          'Could not spawn isolate. Ensure that your file is in a valid directory (i.e. "lib", "web", "test).');
```

Comment:
The general recommended directories for vm apps are `bin`, `test`, `tool`, `benchmark`, `example`, but not `lib`. Basically, anything _but_ `lib` ;). Flutter violates this convention but we don't want to encourage it as standard practice.